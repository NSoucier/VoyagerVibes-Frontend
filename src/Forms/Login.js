import React, { useState, useEffect } from "react";
import { Form, FormGroup, Label, Input, Button } from "reactstrap";
import { Link, useNavigate } from "react-router-dom";
import './UserForms.css' 

function Login({ login, setCurrentUser }) {
    const [ form, setForm ] = useState({username: '', password: ''});
    const [ errorMsg, setErrorMsg ] = useState('');
    const navigate = useNavigate();

    useEffect(() => {
        // redirect to profile if loggedin user requests login page
        if (localStorage.user) navigate('/profile')
    }, []);

    // handles user input to login form
    function handleInput(evt) {
        evt.preventDefault();
        form[evt.target.name] = evt.target.value;
        setForm(form);
    }

    // authenticates user's login credentials
    async function handleClick (evt) {
        evt.preventDefault();

        if(!form.username.length || !form.password.length) {
            setErrorMsg("Please enter username and password.");
        } else {
            // show loading spinner
            const spinner = document.getElementById('buffering');
            spinner.style.display = 'block';

            // check credentials with db
            let status = await login(form);
            // if present, show errors
            if (!status.username) {
                console.error('Error: ', status)
                setErrorMsg(status);
                spinner.style.display = 'none';
            } else {
                // set currentuser state
                setCurrentUser(status.username)
                localStorage.setItem('user', status.username);
                // redirect to explore page upon successful attempt
                navigate('/');
            }            
        }
    }

    return (
        <>
            { !localStorage.user ? (
                <Form id="loginForm">
                    <FormGroup floating>
                        <Input name="username" type="text" placeholder="" onChange={handleInput}/>
                        <Label for="username">Username</Label>
                    </FormGroup>{" "}
                    <FormGroup floating>
                        <Input name="password" type="password" placeholder="" onChange={handleInput}/>
                        <Label for="password">Password</Label>
                    </FormGroup>{" "}
                    <p id="error" style={{color: 'red'}}>{errorMsg}</p>
                    <Button color="success" onClick={handleClick}>Submit</Button>
                    <p id="message">Don't have an account yet? Sign up <Link to='/signup'>here</Link>!</p>
                    <div id="buffering" style={{display: "none"}}>
                        <img src="https://www.jsfirm.com/resources/ajax-loader.gif" id="spinner"/>
                    </div>
                </Form>

            ) : (
                <p>Invalid request.</p>
            )}
        </>     

    );
}

export default Login;
