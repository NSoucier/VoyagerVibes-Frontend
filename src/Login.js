import React, { useState, useEffect } from "react";
import { Form, FormGroup, Label, Input, Button } from "reactstrap";
import { Link, useNavigate } from "react-router-dom";
import './UserForms.css' 

function Login({ login, setCurrentUser }) {
    const [ form, setForm ] = useState({username: '', password: ''});
    const navigate = useNavigate();

    useEffect(() => {
        // redirect to profile if loggedin user requests login page
        if (localStorage.user) navigate('/profile')
    }, []);

    // handles user input to login form
    function handleInput(evt) {
        evt.preventDefault();
        form[evt.target.name] = evt.target.value;
        setForm(form);
    }

    // authenticates user's login credentials
    async function handleClick (evt) {
        evt.preventDefault();
        // show loading spinner
        const spinner = document.getElementById('buffering');
        spinner.innerHTML = '<img src="https://www.jsfirm.com/resources/ajax-loader.gif" id="spinner"/>';
        // check credentials with db
        let status = await login(form);
        // if present, show errors
        if (!status.username) {
            const error = document.getElementById('error');
            error.innerText = status;
            spinner.innerHTML = '';
        } else {
            // set currentuser state
            setCurrentUser(status.username)
            localStorage.setItem('user', status.username);
            // redirect to explore page upon successful attempt
            navigate('/');
        }
    }

    return (
        <>
            { !localStorage.user ? (
                <Form id="loginForm">
                    <FormGroup floating>
                        <Input name="username" type="text" placeholder="" onChange={handleInput}/>
                        <Label for="username">Username</Label>
                    </FormGroup>{" "}
                    <FormGroup floating>
                        <Input name="password" type="password" placeholder="" onChange={handleInput}/>
                        <Label for="password">Password</Label>
                    </FormGroup>{" "}
                    <p id="error" style={{color: 'red'}}></p>
                    <Button color="success" onClick={handleClick}>Submit</Button>
                    <p id="message">Don't have an account yet? Sign up <Link to='/signup'>here</Link>!</p>
                    <div id="buffering"></div>
                </Form>

            ) : (
                <p>Invalid request.</p>
            )}
        </>     

    );
}

export default Login;
