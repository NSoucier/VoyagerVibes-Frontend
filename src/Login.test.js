import React from "react";
import { act } from "react";
import { render, fireEvent } from "@testing-library/react";
import Login from "./Login";
import { Route, Routes, MemoryRouter, BrowserRouter } from "react-router-dom";

// jest.setup.js
const originalConsoleWarn = console.warn;

beforeEach(() => {
    console.log('mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm')
  console.warn = (...args) => {
    if (args[0] && args[0].includes('ReactDOMTestUtils.act is deprecated')) {
      return;
    }
    originalConsoleWarn(...args);
  };
});

afterEach(() => {
  console.warn = originalConsoleWarn;
});

// smoke test
it("renders without crashing", async function () {
    act(() => {
        render(
            <BrowserRouter>
                <Login />
            </BrowserRouter>
          );
    });
});

// snapshot test
// it("matches snapshot", function() {
//     const {asFragment} = render(
//         <BrowserRouter>
//             <Login />
//         </BrowserRouter>
//       );
//     expect(asFragment()).toMatchSnapshot();
// });