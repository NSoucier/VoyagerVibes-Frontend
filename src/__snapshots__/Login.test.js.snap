// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
    id="loginForm"
  >
    <div
      class="mb-3 form-floating"
    >
      <input
        class="form-control"
        name="username"
        placeholder=""
        type="text"
      />
      <label
        class="form-label"
        for="username"
      >
        Username
      </label>
    </div>
     
    <div
      class="mb-3 form-floating"
    >
      <input
        class="form-control"
        name="password"
        placeholder=""
        type="password"
      />
      <label
        class="form-label"
        for="password"
      >
        Password
      </label>
    </div>
     
    <p
      id="error"
      style="color: red;"
    />
    <button
      class="btn btn-success"
      type="button"
    >
      Submit
    </button>
    <p
      id="message"
    >
      Don't have an account yet? Sign up 
      <a
        href="/signup"
      >
        here
      </a>
      !
    </p>
  </form>
</DocumentFragment>
`;
